Change tmpDir when running e2e localmachine test on macOS

tnk4on opened the pull request:
Fix https://github.com/containers/podman/issues/22360

#### Does this PR introduce a user-facing change?

```release-note
Fixed an error starting Podman machine when running e2e localmachine tests on macOS.
```

The fixes are only to the e2e test file and Doc, no additional tests.

ashley-cui reviewed (commented):
Thanks @tnk4on!  
I believe our macOS CI sets `TMPDIR`, so it makes sense that the tests are failing. I think instead, we should only set tmpdir on macOS to `/private/tmp` if `TMPDIR` is not set, rather than overriding it.

tnk4on commented:
@ashley-cui 
I think the macOS CI Fail is another cause as it was happening with other PRs. The podman machine actually started without timeout until halfway through.

I added code to check $TMPDIR. Please check.

ashley-cui commented on pkg/machine/e2e/machine_test.go:
```diff
@@ -169,3 +172,16 @@ func copySparse(dst io.WriteSeeker, src io.Reader) (retErr error) {
+
+func getTmpDir(value string) string {
+	if runtime.GOOS == "darwin" {
+		if len(value) >= 22 || value == "" {
```
I think I would still prefer that we only change the tmpdir if it is not set. If the user/runner of tests sets a path len > 22, I think we should fail with an error, and not secretly change the path without notice. 

macOS tests are passing now!

tnk4on commented:
@ashley-cui I added some code. Please check.

If TMPDIR=“”, it cannot connect to gvproxy.sock.
```
  > Enter [It] Volume ops - /Users/shtanaka/dev2/podman/pkg/machine/e2e/basic_test.go:73 @ 05/01/24 09:53:43.085
  /Users/shtanaka/dev2/podman/bin/darwin/podman machine init --disk-size 11 --image /private/tmp/podman_test234588004/podman-machine-daily.aarch64.applehv.raw --now 7f8b5007b0ea
  Machine init complete
  Starting machine "7f8b5007b0ea"
  Error: unable to connect to "gvproxy" socket at "podman/7f8b5007b0ea-gvproxy.sock"
  [FAILED] Expected
      <int>: 125
  to match exit code:
      <int>: 0
  In [It] at: /Users/shtanaka/dev2/podman/pkg/machine/e2e/basic_test.go:95 @ 05/01/24 09:54:26.226
  ```
  so I changed it to return an error.
  
  Also, if TMPDIR is a directory that is not mounted on the podman machine, the bind mount test (e.g. "Volume ops". at basic_test.go) will return an error.
Therefore, added check if TMPDIR is included in Machine.Volumes.

ashley-cui commented on pkg/machine/e2e/machine_test.go:
```diff
@@ -169,3 +183,34 @@ func copySparse(dst io.WriteSeeker, src io.Reader) (retErr error) {
+	case runtime.GOOS != "darwin":
+		tmpDir = value
+	case len(value) >= 22:
+		return "", errors.New("path length should be less than 22 characters")
```
Nit, so the user knows what path it is: 
```suggestion
		return "", errors.New("$TMPDIR path length should be less than 22 characters")
```

tnk4on replied:
small change and fixed

ashley-cui commented on pkg/machine/e2e/machine_test.go:
```diff
@@ -169,3 +183,34 @@ func copySparse(dst io.WriteSeeker, src io.Reader) (retErr error) {
+	case len(value) >= 22:
+		return "", errors.New("path length should be less than 22 characters")
+	case value == "":
+		return "", errors.New("TMPDIR cannot be empty. Set to directory mounted on podman machine (e.g. /private/tmp)")
```
I'm okay with setting TMPDIR automatically to `/private/tmp` in this instance

tnk4on replied:
This seems to be another Issue.
```
% export TMPDIR=""
% podman machine init --now
Looking up Podman Machine image at quay.io/podman/machine-os:5.1 to create VM
Getting image source signatures
Copying blob 5ab0b6fcee8a done   | 
Copying config 44136fa355 done   | 
Writing manifest to image destination
5ab0b6fcee8aba3f303ad47032c77ffa6894a25f8bc3645b3567860cd7dddf39
Extracting compressed file: podman-machine-default-arm64.raw: done  
Machine init complete
Starting machine "podman-machine-default"
Error: unable to connect to "gvproxy" socket at "podman/podman-machine-default-gvproxy.sock"
```

The following code will set an empty path if `TMPDIR=“”`. So the gvproxy path will be inappropriate.
If `unset TMPDIR` is executed, /tmp is set and it works.

https://github.com/containers/podman/blob/c9644ebccf14309a77769cba00833cd139509e4a/pkg/machine/env/dir_darwin.go#L5-L11

ashley-cui reviewed (commented):
Other than two small comments, LGTM. Thanks!

ashley-cui commented:
LGTM @containers/podman-maintainers  PTAL
